File: lib/bid-strategy.ts
/**
 * Bid pricing strategy.
 *
 * Turns the accounting-derived Total Project Cost into a recommended bid
 * price by applying a risk-adjusted margin: the margin is the lever that
 * trades off win probability (lower price wins more often) against
 * profitability (higher price earns more per win). Deterministic - no LLM.
 */

export type RiskLevel = 'LOW' | 'MEDIUM' | 'HIGH';

export interface BidStrategyResult {
  total_project_cost: number;
  margin_percentage: number; // fraction, e.g. 0.12 = 12%
  recommended_bid_price: number;
  profit_amount: number;
  rationale: string;
}

/**
 * Risk-adjusted margin tiers.
 *
 * LOW risk gets the tightest margin: the project is safe enough that
 * pricing aggressively (still profitable) maximizes the chance of winning.
 * MEDIUM risk (PROCEED_WITH_CAUTION) carries a heavier margin to compensate
 * for the added exposure, even though it slightly reduces win odds.
 * HIGH risk is priced defensively - this tier should generally correspond
 * to a "do not proceed" recommendation, so the price here is a walk-away
 * reference point rather than a competitive bid.
 */
const MARGIN_BY_RISK_LEVEL: Record<RiskLevel, number> = {
  LOW: 0.12,
  MEDIUM: 0.18,
  HIGH: 0.25,
};

function round2(value: number): number {
  return Math.round(value * 100) / 100;
}

export function calculateOptimalBidPrice(
  totalProjectCost: number,
  riskLevel: RiskLevel,
): BidStrategyResult {
  const margin_percentage = MARGIN_BY_RISK_LEVEL[riskLevel] ?? MARGIN_BY_RISK_LEVEL.MEDIUM;
  const recommended_bid_price = round2(totalProjectCost * (1 + margin_percentage));
  const profit_amount = round2(recommended_bid_price - totalProjectCost);

  const rationale =
    riskLevel === 'LOW'
      ? `Priced at a ${(margin_percentage * 100).toFixed(0)}% margin over the $${totalProjectCost.toLocaleString()} project cost. Risk is low, so a tighter margin keeps this bid competitive while remaining solidly profitable.`
      : riskLevel === 'MEDIUM'
        ? `Priced at a ${(margin_percentage * 100).toFixed(0)}% margin over the $${totalProjectCost.toLocaleString()} project cost. The added margin compensates for moderate risk exposure while still leaving room to win the bid.`
        : `Priced at a ${(margin_percentage * 100).toFixed(0)}% margin over the $${totalProjectCost.toLocaleString()} project cost. This is a defensive, walk-away price - risk is high enough that winning at a lower price would not be worth the exposure.`;

  return {
    total_project_cost: totalProjectCost,
    margin_percentage,
    recommended_bid_price,
    profit_amount,
    rationale,
  };
}

File: python/app/bid_strategy.py
"""Risk-adjusted bid pricing strategy. Ported from lib/bid-strategy.ts."""

from __future__ import annotations

from typing import Literal

RiskLevel = Literal["LOW", "MEDIUM", "HIGH"]

MARGIN_BY_RISK_LEVEL: dict[RiskLevel, float] = {
    "LOW": 0.12,
    "MEDIUM": 0.18,
    "HIGH": 0.25,
}


def _round2(value: float) -> float:
    return round(value * 100) / 100


def calculate_optimal_bid_price(total_project_cost: float, risk_level: RiskLevel) -> dict:
    margin_percentage = MARGIN_BY_RISK_LEVEL.get(risk_level, MARGIN_BY_RISK_LEVEL["MEDIUM"])
    recommended_bid_price = _round2(total_project_cost * (1 + margin_percentage))
    profit_amount = _round2(recommended_bid_price - total_project_cost)

    cost_str = f"{total_project_cost:,.0f}"
    margin_str = f"{margin_percentage * 100:.0f}"

    if risk_level == "LOW":
        rationale = (
            f"Priced at a {margin_str}% margin over the ${cost_str} project cost. "
            "Risk is low, so a tighter margin keeps this bid competitive while remaining solidly profitable."
        )
    elif risk_level == "MEDIUM":
        rationale = (
            f"Priced at a {margin_str}% margin over the ${cost_str} project cost. "
            "The added margin compensates for moderate risk exposure while still leaving room to win the bid."
        )
    else:
        rationale = (
            f"Priced at a {margin_str}% margin over the ${cost_str} project cost. "
            "This is a defensive, walk-away price - risk is high enough that winning at a lower price "
            "would not be worth the exposure."
        )

    return {
        "total_project_cost": total_project_cost,
        "margin_percentage": margin_percentage,
        "recommended_bid_price": recommended_bid_price,
        "profit_amount": profit_amount,
        "rationale": rationale,
    }
